File: smoke_tests/php/denial_of_service.php
<?php
// Denial of Service vulnerabilities in PHP

// Test 1: Unbounded array allocation
function create_array() {
    $size = (int)$_GET['size'];
    // VULNERABLE: User controls array size
    $data = array_fill(0, $size, 'x');
    echo json_encode(['length' => count($data)]);
}

// Test 2: Unbounded string multiplication
function repeat_string() {
    $count = (int)$_GET['count'];
    // VULNERABLE: User controls repetition
    $result = str_repeat('x', $count);
    echo $result;
}

// Test 3: ReDoS
function validate_input() {
    $input = $_GET['input'];
    // VULNERABLE: Catastrophic backtracking
    preg_match('/^(a+)+$/', $input);
}

// Test 4: XML bomb (billion laughs)
function parse_xml() {
    $xml = $_POST['xml'];
    // VULNERABLE: Entity expansion not limited
    libxml_disable_entity_loader(false);
    $doc = new DOMDocument();
    $doc->loadXML($xml, LIBXML_NOENT);
    echo $doc->saveXML();
}

// Test 5: Hash collision
function store_data() {
    // VULNERABLE: Hash collision with crafted keys
    $data = [];
    foreach ($_POST as $key => $value) {
        $data[$key] = $value;
    }
    echo json_encode(['count' => count($data)]);
}

// Test 6: JSON parsing large numbers
function parse_json() {
    $json = $_POST['json'];
    // VULNERABLE: Large numbers/deep nesting
    $data = json_decode($json, true, 512);
    echo json_encode($data);
}

// Test 7: CPU exhaustion
function compute() {
    $iterations = (int)$_GET['n'];
    // VULNERABLE: User controls computation
    $result = 0;
    for ($i = 0; $i < $iterations; $i++) {
        $result += sin($i) * cos($i);
    }
    echo $result;
}

// Test 8: File read amplification
function read_file() {
    $path = $_GET['path'];
    // VULNERABLE: No size limit on file read
    echo file_get_contents($path);
}

// Test 9: ZIP bomb
function extract_zip() {
    $file = $_FILES['file'];
    $zip = new ZipArchive();
    $zip->open($file['tmp_name']);
    // VULNERABLE: No decompression ratio limit
    $zip->extractTo('/tmp/extracted/');
    $zip->close();
}

// Test 10: Synchronous blocking
function slow_operation() {
    $delay = (int)$_GET['delay'];
    // VULNERABLE: User controls blocking time
    sleep($delay);
    echo 'done';
}

// Test 11: Fork bomb setup
function spawn_processes() {
    $count = (int)$_GET['count'];
    // VULNERABLE: User controls process spawning
    for ($i = 0; $i < $count; $i++) {
        pcntl_fork();
    }
}

// Test 12: Memory exhaustion via image
function process_image() {
    $width = (int)$_GET['width'];
    $height = (int)$_GET['height'];
    // VULNERABLE: User controls image dimensions
    $image = imagecreatetruecolor($width, $height);
    // Large dimensions cause memory exhaustion
}

// Test 13: Recursive function with user depth
function recursive_process($depth = 0) {
    $max_depth = (int)$_GET['depth'];
    // VULNERABLE: User controls recursion depth
    if ($depth < $max_depth) {
        recursive_process($depth + 1);
    }
}

// Test 14: File upload exhaustion
function handle_uploads() {
    // VULNERABLE: No limit on number of files
    foreach ($_FILES as $file) {
        move_uploaded_file($file['tmp_name'], '/uploads/' . $file['name']);
    }
}
?>

File: smoke_tests/ruby/denial_of_service.rb
# Denial of Service vulnerabilities in Ruby/Rails
require 'json'
require 'nokogiri'
require 'yaml'

class DosController < ApplicationController
  # Test 1: Unbounded array allocation
  def create_array
    size = params[:size].to_i
    # VULNERABLE: User controls array size
    data = Array.new(size)
    render json: { length: data.length }
  end

  # Test 2: Unbounded string multiplication
  def repeat_string
    count = params[:count].to_i
    # VULNERABLE: User controls repetition
    result = 'x' * count
    render plain: result
  end

  # Test 3: ReDoS
  def validate_input
    input = params[:input]
    # VULNERABLE: Catastrophic backtracking
    pattern = /^(a+)+$/
    render json: { match: pattern.match?(input) }
  end

  # Test 4: XML bomb
  def parse_xml
    xml = params[:xml]
    # VULNERABLE: Entity expansion not limited
    doc = Nokogiri::XML(xml) do |config|
      config.noent  # Enables entity substitution
    end
    render plain: doc.to_xml
  end

  # Test 5: YAML deserialization bomb
  def parse_yaml
    yaml = params[:yaml]
    # VULNERABLE: Can contain recursive structures
    data = YAML.unsafe_load(yaml)
    render json: data
  end

  # Test 6: JSON parsing large numbers
  def parse_json
    json = params[:json]
    # VULNERABLE: Large numbers can cause DoS
    data = JSON.parse(json)
    render json: data
  end

  # Test 7: CPU exhaustion
  def compute
    iterations = params[:n].to_i
    # VULNERABLE: User controls computation
    result = 0
    iterations.times { |i| result += Math.sin(i) * Math.cos(i) }
    render json: { result: result }
  end

  # Test 8: File read amplification
  def read_file
    path = params[:path]
    # VULNERABLE: No size limit on file read
    content = File.read(path)
    render plain: content
  end

  # Test 9: Hash collision attack
  def store_data
    # VULNERABLE: Hash collision with crafted keys
    data = {}
    params.each { |k, v| data[k] = v }
    render json: { count: data.size }
  end

  # Test 10: Recursive structure
  def process_nested
    data = params[:data]
    # VULNERABLE: Deep nesting can cause stack overflow
    def deep_process(obj, depth = 0)
      return obj if depth > 1000
      case obj
      when Hash
        obj.transform_values { |v| deep_process(v, depth + 1) }
      when Array
        obj.map { |v| deep_process(v, depth + 1) }
      else
        obj
      end
    end
    result = deep_process(data)
    render json: result
  end

  # Test 11: Synchronous blocking
  def slow_operation
    delay = params[:delay].to_i
    # VULNERABLE: User controls blocking time
    sleep(delay)
    head :ok
  end

  # Test 12: Unbounded loop
  def process_items
    count = params[:count].to_i
    # VULNERABLE: User controls iteration count
    count.times do
      # Do work
      Object.new
    end
    head :ok
  end

  # Test 13: ZIP bomb
  def extract_zip
    file = params[:file]
    temp_path = Rails.root.join('tmp', file.original_filename)
    File.binwrite(temp_path, file.read)
    # VULNERABLE: No decompression ratio limit
    Zip::File.open(temp_path) do |zip_file|
      zip_file.each do |entry|
        entry.extract(Rails.root.join('extracted', entry.name))
      end
    end
    head :ok
  end
end
